src/log-store/src/kafka/producer.rs

```diff
@@ -0,0 +1,561 @@
+        if let Some(flush_token) = flush_token {
+            if self.flush_clock != flush_token {
+                // The flush has been triggered.
+                return Ok(());
+            }
+        }
+
+        let maybe_flush_request = self.batch_builder.yield_flush();
+        self.has_leader = false;
+        self.flush_clock = self.flush_clock.wrapping_add(1);
+
+        match maybe_flush_request {
+            Ok(Some(flush_request)) => {
+                self.sender.send(flush_request).await.map_err(|err| {
+                    error::SendFlushRequestSnafu {
+                        reason: err.to_string(),
+                    }
+                    .build()
+                })?;
+            }
+            Ok(None) => {
+                // Nothing to flush
+            }
+            Err(err) => return Err(err),
+        }
+
+        Ok(())
+    }
+}
+
+/// [`OrderedBatchProducer`] attempts to aggregate multiple produce requests together
+/// using the provided [`Aggregator`].
+///
+/// It will buffer up records until either the linger time expires or
+/// [`Aggregator`] cannot accommodate another record.
+///
+/// At this point it will flush the [`Aggregator`].
+///
+/// The flush will respect the order of aggregated data.
+pub struct OrderedBatchProducer<A: Aggregator> {
+    /// The timeout of performing aggregating.
+    linger: Duration,
+    inner: Arc<Mutex<OrderedBatchProducerInner<A>>>,
+    /// Used to control the [`BackgroundProducerWorker`].
+    running: Arc<AtomicBool>,
+    /// The handle of [`BackgroundProducerWorker`].
+    handle: JoinHandle<()>,
+}
+
+impl<A: Aggregator> Drop for OrderedBatchProducer<A> {
+    fn drop(&mut self) {
+        self.running.store(false, Ordering::Relaxed);
+    }
+}
+
+impl<A: Aggregator> OrderedBatchProducer<A> {
+    /// Constructs a new [`OrderedBatchProducer`].
+    pub(crate) fn new(
+        aggregator: A,
+        client: Arc<dyn ProducerClient>,
+        linger: Duration,
+        compression: Compression,
+        flush_queue_size: usize,
+    ) -> Self {
+        let (tx, rx) = mpsc::channel(flush_queue_size);
+        let running = Arc::new(AtomicBool::new(true));
+        let mut worker = BackgroundProducerWorker {
+            client,
+            receiver: rx,
+            compression,
+            running: running.clone(),
+            flushing: None,
+        };
+
+        Self {
+            linger,
+            inner: Arc::new(Mutex::new(OrderedBatchProducerInner::new(aggregator, tx))),
+            running,
+            handle: tokio::spawn(async move { worker.run().await }),
+        }
+    }
+
+    /// Writes `data` to the [`OrderedBatchProducer`].
+    ///
+    /// Returns the [ResultReceiver], which will receive a result when data has been committed to Kafka
+    /// or an unrecoverable error has been encountered.
+    pub(crate) async fn produce(&self, data: A::Input) -> Result<ResultReceiver<A>> {
```

Comment:
If we can support producing multiple inputs at once (i.e., `Vec<Input>`), it might be beneficial for improving throughput.